Fix error message

Zastai opened the pull request:
### Proposed changelog entries

(none - too trivial)

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

Zastai committed:
Fix error message

daniel-beck commented on core/src/main/java/hudson/console/ConsoleLogFilter.java:
```diff
@@ -69,7 +69,7 @@ public OutputStream decorateLogger(AbstractBuild build, OutputStream logger) thr
             // happens only if the subtype fails to override neither decorateLogger method
```
```suggestion
            // happens only if the subtype fails to override either decorateLogger method
```

daniel-beck commented on core/src/main/java/hudson/console/ConsoleLogFilter.java:
```diff
@@ -69,7 +69,7 @@ public OutputStream decorateLogger(AbstractBuild build, OutputStream logger) thr
             throw new AssertionError("The plugin '" + this.getClass().getName() + "' still uses " +
                     "deprecated decorateLogger(AbstractBuild,OutputStream) method. " +
-                    "Update the plugin to use setUp(Run,OutputStream) instead.");
+                    "Update the plugin to use decorateLogger(Run,OutputStream) instead.");
```
```suggestion
                    "Update the plugin to call #decorateLogger(Run,OutputStream) instead.");
```

It was unclear to me what is meant here because I first thought it was referring to what was implemented (and that seemed to contradict the comment above). "Call" would be clearer.

OTOH, why would you report the current class's name if you're interested in the caller of a public method? Something's weird here.

Zastai replied:
Not sure I'd add the `#` because this isn't JavaDoc, but not against it either.
Adding the class name would be OK, but it would be reported next to the message in the exception's stack trace too.

As for the rest of the wording, there are many similar pieces of code similar to this (hence the copy/paste error of the method name in the first place), so if that's getting changed, it should probably be changed across the board.

Zastai replied:
Actually, that branch is reached when a `ConsoleFilter` overrides neither method, which is logically an `AbstractMethodError()`.
So if anything, it should be "implemenet" or "override" instead of "use" or "call".
Or just drop the `AssertionError` and throw `AbstractMethodException` instead, which is what `SimpleBuildStep` does.

Whatever change is made should then also be made in `BuildWrapper` (looks like it's where this was copied from).

daniel-beck replied:
> Actually, that branch is reached when a `ConsoleFilter` overrides neither method, which is logically an `AbstractMethodError()`.

Right, but then the message is just plain wrong. The mentioned method isn't overridden either.

Zastai replied:
I've changed both places to use `AbstractMethodError` with a better message.

daniel-beck reviewed (commented):
The immediate and obvious problem of the bad method name is being fixed, but it seems to me that there's something else going on here. I hesitate to approve, not because there's something wrong with this contribution, but because we should figure out whether this needs a slightly larger change to be _really_ correct, and implement that.

Zastai committed:
Fix English-o

Co-authored-by: Daniel Beck <1831569+daniel-beck@users.noreply.github.com>

Zastai committed:
Use AbstractMethodError

For both ConsoleLogFilter and BuildWrapper, to better describe what the
problem is.

MRamonLeon commented on core/src/main/java/hudson/console/ConsoleLogFilter.java:
```diff
@@ -66,10 +66,9 @@ public OutputStream decorateLogger(AbstractBuild build, OutputStream logger) thr
-                    "Update the plugin to use setUp(Run,OutputStream) instead.");
+            // happens only if the subtype fails to override either decorateLogger method
+            throw new AbstractMethodError("Plugin class '" + this.getClass().getName() + "' does not override " +
+                    "either overload of the decorateLogger method.");
```
```suggestion
                    "either overload the decorateLogger method.");
```

daniel-beck replied:
Wrong, `overload` here is an noun.

MRamonLeon commented on core/src/main/java/hudson/tasks/BuildWrapper.java:
```diff
@@ -140,10 +140,9 @@ public boolean tearDown( Build build, BuildListener listener ) throws IOExceptio
-                    "Update the plugin to use setUp(AbstractBuild, Launcher, BuildListener) instead.");
+        else // neither overload was overridden
+            throw new AbstractMethodError("Plugin class '" + this.getClass().getName() + "' does not override " +
+                    "either overload of the setUp method.");
```
```suggestion
                    "either overload the setUp method.");
```

daniel-beck replied:
Wrong, `overload` here is an noun.

MRamonLeon commented on core/src/main/java/hudson/tasks/BuildWrapper.java:
```diff
@@ -140,10 +140,9 @@ public boolean tearDown( Build build, BuildListener listener ) throws IOExceptio
-        else
-            throw new AssertionError("The plugin '" + this.getClass().getName() + "' still uses " +
-                    "deprecated setUp(Build,Launcher,BuildListener) method. " +
-                    "Update the plugin to use setUp(AbstractBuild, Launcher, BuildListener) instead.");
```
don't we want this information?

daniel-beck replied:
As above.

MRamonLeon commented on core/src/main/java/hudson/console/ConsoleLogFilter.java:
```diff
@@ -66,10 +66,9 @@ public OutputStream decorateLogger(AbstractBuild build, OutputStream logger) thr
-            // happens only if the subtype fails to override neither decorateLogger method
-            throw new AssertionError("The plugin '" + this.getClass().getName() + "' still uses " +
-                    "deprecated decorateLogger(AbstractBuild,OutputStream) method. " +
-                    "Update the plugin to use setUp(Run,OutputStream) instead.");
```
don't we want this information?

daniel-beck replied:
It's wrong.

daniel-beck reviewed (approved):
Seems reasonable although I would not be surprised to learn there's something weird going on and this is wrong after all 😄 Thanks for the update.

oleg-nenashev commented on core/src/main/java/hudson/tasks/BuildWrapper.java:
```diff
@@ -140,10 +140,9 @@ public boolean tearDown( Build build, BuildListener listener ) throws IOExceptio
-                    "deprecated setUp(Build,Launcher,BuildListener) method. " +
-                    "Update the plugin to use setUp(AbstractBuild, Launcher, BuildListener) instead.");
+        else // neither overload was overridden
+            throw new AbstractMethodError("Plugin class '" + this.getClass().getName() + "' does not override " +
```
Add to Javadoc?

Zastai replied:
Done.

oleg-nenashev reviewed (approved):
Looks good to me. `AbstractMethodError` has a slightly different purpose, but it is definitely better than the current exception.

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

oleg-nenashev commented:
i would recommend adding it to the changelog as a developer change

Zastai commented:
Could do, but the change is essentially "if your plugin did not override at least one overload, it will now produce a different failure than before". In practice there should not be any failures as it stands, so that the failure will look different should not be all that relevant.

oleg-nenashev commented:
OK, let's keep it as is then

Zastai committed:
Document the exception

Zastai committed:
Adjust the error handling for BuildWrapper.setUp

This adds an override check to the deprecated overload, moving the
AbstractMethodError there to match.
The error in the new API is now an UnsupportedOperationException
because it's a case of an old plugin (supporting only Build) being
called for a non-Build build (e.g. MavenBuild).

Zastai committed:
Correct class name

Zastai commented:
I noticed the `BuildWrapper` case was slightly different, making the error message less accurate. I adjusted it to match the other code.

This means it has two failure cases now:
1. a plugin implements only the old API (taking `Build`) but the method is called for another type of `AbstractBuild`
   - now results in `UnsupportedOperationException`
2. a plugin implements neither overload
   - results in `AbstractMethodError` when the old API is called

oleg-nenashev commented:
needs re-review after the changes.

oleg-nenashev reviewed (approved):
Looks good to me after the re-review.

oleg-nenashev commented:
We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process
